src/app/space/settings/areas/areas.component.ts

```diff
@@ -18,69 +30,251 @@ export class AreasComponent implements OnInit, OnDestroy {
+
+  compare(area1: Area, area2: Area): number {
+    var compValue = 0;
+    if (this.currentSortField === undefined || this.currentSortField.id === 'area') {
+      compValue = area1.attributes.name.localeCompare(area2.attributes.name);
+    }
+    if (!this.isAscendingSort) {
+      compValue = compValue * -1;
+    }
+    return compValue;
+  }
+
+  sortChange($event: SortEvent): void {
+    this.currentSortField = $event.field;
+    this.isAscendingSort = $event.isAscending;
+    this.allAreas.sort((area1: Area, area2: Area) => this.compare(area1, area2));
+    this.applyFilters(this.appliedFilters); // Reapply filters, if any
+  }
+
+  // Transform flat array into tree list
+
+  private buildTree(elements: any[], tree = []): any[] {
+    elements.forEach((element) => {
+      if (element.relationships.parent === undefined) {
+        let children = this.getNestedChildren(elements, element);
+        if (children.length > 0) {
+          element.children = children;
+          element.expanded = true;
+        }
+        tree.push(element);
+      }
+    });
+    return tree;
+  }
+
+  private getNestedChildren(elements: any[], parent: Area): any[] {
+    let areas = [];
+    elements.forEach((element) => {
+      if (element.relationships.parent !== undefined && element.relationships.parent.data.id === parent.id) {
+        let children = this.getNestedChildren(elements, element);
+        if (children.length > 0) {
+          element.children = children;
+          element.expanded = true;
+        }
+        areas.push(element);
+      }
+    });
+    return areas;
+  }
+
+  // Transform filtered flat array into tree list
+
+  private buildFilteredTree(elements: any[]): any[] {
+    // Reassign filtered parents using closest ancestor
+    elements.forEach((element) => {
+      element.children = undefined;
+      if (element.relationships.parent !== undefined) {
+        let area = this.getClosestAncestor(elements,
+          this.getArea(element.relationships.parent.data.id));
+        if (area !== undefined) {
+          element.relationships.parent.data = area;
+        } else {
+          element.relationships.parent = undefined;
+        }
+      }
+    });
+    return this.buildTree(elements);
+  }
+
+  private getClosestAncestor(elements: any[], parent: Area): Area {
+    let area;
+    for (let i = 0; i < elements.length; i++) {
+      if (elements[i].id === parent.id) {
+        area = parent;
+        break;
+      }
+    }
+    if (area === undefined && parent.relationships.parent !== undefined) {
+      area = this.getClosestAncestor(elements,
+        this.getArea(parent.relationships.parent.data.id));
+    }
+    return area;
+  }
+
+  private getArea(id: string): Area {
+    let area;
+    for (let i = 0; i < this.allAreas.length; i++) {
+      if (this.allAreas[i].id === id) {
+        area = this.allAreas[i];
+      }
+    }
+    return area;
+  }
+
+  // Actions
+
+  handleAction($event: Action, node: any): void {
```

Comment:
Is there a type for the `node` parameter? The function's implementation assumes it has specific properties while the signature suggests it can take any kind of object.

Reply:
Unfortunately, this object is from the underlying angular-tree-component, which is type any.